webcrypto/subtle_crypto.go

```diff
@@ -137,8 +141,87 @@ func (sc *SubtleCrypto) Verify(
 //   - SHA-512
 //
 // The `data` parameter should contain the data to be digested.
-func (sc *SubtleCrypto) Digest(algorithm goja.Value, data interface{}) *goja.Promise {
-	return nil
+func (sc *SubtleCrypto) Digest(algorithm interface{}, data goja.Value) *goja.Promise {
+	promise, resolve, reject := sc.makeHandledPromise()
+
+	// Validate that the value we received is either an ArrayBuffer, TypedArray, or DataView
+	// This uses the technique described in https://github.com/dop251/goja/issues/379#issuecomment-1164441879
+	var (
+		isArrayBuffer = IsInstanceOf(sc.vu.Runtime(), data, ArrayBufferConstructor)
+		isDataView    = IsInstanceOf(sc.vu.Runtime(), data, DataViewConstructor)
+		isTypedArray  = IsTypedArray(sc.vu.Runtime(), data)
+	)
+
+	if !isArrayBuffer && !isDataView && !isTypedArray {
+		reject(fmt.Errorf("data must be an ArrayBuffer, TypedArray, or DataView"))
+		return promise
+	}
+
+	// 2.
+	// Cast the data to a Goja Object, and, as we're now sure it's
+	// either an ArrayBuffer, or a view on an ArrayBuffer, we can
+	// get the underlying ArrayBuffer by exporting its buffer property
+	// to a Goja ArrayBuffer, and then getting its underlying Go slice
+	// by calling the `Bytes()` method.
+	//
+	// Doing so conviniently also copies the underlying buffer, which
+	// is required by the specification.
+	// See https://www.w3.org/TR/WebCryptoAPI/#SubtleCrypto-method-digest
+	asObject := data.ToObject(sc.vu.Runtime())
+	arrayBuffer, ok := asObject.Export().(goja.ArrayBuffer)
+	if !ok {
+		reject(fmt.Errorf("could not get ArrayBuffer from data"))
+		return promise
+	}
+	bytes := arrayBuffer.Bytes()
+
+	// The specification explicitly requires us to copy the underlying
+	// bytes held by the array buffer
+	bytesCopy := make([]byte, len(bytes))
+	copy(bytesCopy, bytes)
+
+	// 3.
+	normalizedAlgorithm, err := NormalizeAlgorithm(algorithm, OperationIdentifierDigest)
+	if err != nil {
+		// "if an error occurred, return a Promise rejected with NormalizedAlgorithm"
+		reject(err)
+		return promise
+	}
+
+	alg, ok := normalizedAlgorithm.(Algorithm)
+	if !ok {
+		reject(fmt.Errorf("algorithm must be a string or an object"))
+		return promise
+	}
+
+	// 6.
+	go func() {
+		var hash hash.Hash
+
+		switch alg.Name {
+		case Sha1:
+			hash = crypto.SHA1.New()
+		case Sha256:
+			hash = crypto.SHA256.New()
+		case Sha384:
+			hash = crypto.SHA384.New()
+		case Sha512:
+			hash = crypto.SHA512.New()
+		default:
```

Comment:
In principle, I agree. However, the traits of the types that could be returned by `NormalizeAlgorithm` will vary quite a bit. In certain scenarios, it will need to return something that can be used to generate a key, in other scenarios, something that can be used to export a key or encrypt. Hence, returning an interface might turn out complex, if not impossible. I'm experimenting with it, and I'll get back to you folks 👍🏻

Reply:
I've substantially cleaned the algorithm normalization up. Plus, I added a `NormalizedAlgorithm` interface which allows doing exactly what you requested. I'm not sure it will survive the "fire baptism" in larger context (`generateKey`, `encrypt`, etc), but the design is already better like this, and it will force us to find better designs down the road indeed 🙇🏻 👍🏻